lib/models/addon.js

```diff
@@ -29,13 +31,59 @@ var ensurePosixPath = require('ensure-posix-path');
 var defaultsDeep = require('ember-cli-lodash-subset').defaultsDeep;
 var Babel = require('broccoli-babel-transpiler');
 var findAddonByName = require('../utilities/find-addon-by-name');
+var experiments = require('../experiments');
+var heimdall = require('heimdalljs');
+var calculateCacheKeyForTree = require('../utilities/addon-calculate-cache-key-for-tree');
+
+heimdall.registerMonitor('addon-tree-cache', function AddonTreeCacheSchema() {
+  this.hits = 0;
+  this.misses = 0;
+  this.adds = 0;
+});
 
 var DEFAULT_BABEL_CONFIG = {
   modules: 'amdStrict',
   moduleIds: true,
   resolveModuleSource: require('amd-name-resolver').moduleResolve
 };
 
+var DEFAULT_TREE_FOR_METHODS = {
```

Comment:
Also, I realized I didn't address this concern in my comment above:

> Depending upon when that gets called it doesn't necessarily go through treeFor as a choke point and could invalidate cache strategy.

Yes, you are absolutely correct! If someone does `this.eachAddonInvoke('treeForEngine')`, it completely bypasses our ability to cache.  This is totally fine, they just don't automatically get the benefits of the new system that we have in place to speed things (like engines / fastboot / etc) up.